kong/clustering.lua

```diff
@@ -304,68 +326,117 @@ function _M.handle_cp_websocket()
+        -- queue PONG to avoid races
+        table_insert(queue, "PONG")
+        queue.sem:post()
+
+        last_seen = ngx_time()
+
+        local ok
+        ok, err = kong.db.clustering_data_planes:upsert({ id = node_id, }, {
+          last_seen = last_seen,
+          config_hash = data ~= "" and data or nil,
+          hostname = node_hostname,
+          ip = node_ip,
+        }, { ttl = kong.configuration.cluster_data_plane_purge_delay, })
+        if not ok then
+          ngx_log(ngx_ERR, "unable to update clustering data plane status: ", err)
+        end
       end
     end
   end)
 
-  while not exiting() do
-    local ok, err = sem:wait(5)
-    if exiting() then
-      return
-    end
-    if ok then
-      local payload = table_remove(queue, 1)
-      assert(payload, "config queue can not be empty after semaphore returns")
-
-      if payload == "PONG" then
-        local _
-        _, err = wb:send_pong()
-        if err then
-          ngx_log(ngx_ERR, "failed to send PONG back to data plane: ", err)
-          return ngx_exit(ngx_ERR)
+  local main_thread = ngx.thread.spawn(function()
+    while not exiting() do
+      local ok, err = sem:wait(5)
+      if exiting() then
+        return
+      end
+      if ok then
+        local payload = table_remove(queue, 1)
+        if not payload then
+          return nil, "config queue can not be empty after semaphore returns"
         end
 
-        ngx_log(ngx_DEBUG, "sent PONG packet to data plane")
+        if payload == "PONG" then
+          local _, err = wb:send_pong()
+          if err then
+            if not is_timeout(err) then
+              return nil, "failed to send PONG back to data plane: " .. err
+            end
 
-      else -- config update
-        local _, err = wb:send_binary(payload)
-        if err then
-          ngx_log(ngx_ERR, "unable to send updated configuration to node: ", err)
+            ngx_log(ngx_NOTICE, "failed to send PONG back to data plane: ", err)
 
-        else
-          ngx_log(ngx_DEBUG, "sent config update to node")
+          else
+            ngx_log(ngx_DEBUG, "sent PONG packet to data plane")
+          end
+
+        else -- config update
+          local _, err = wb:send_binary(payload)
+          if err then
+            if not is_timeout(err) then
+              return nil, "unable to send updated configuration to node: " .. err
+            end
+
+            ngx_log(ngx_NOTICE, "unable to send updated configuration to node: ", err)
+
+          else
+            ngx_log(ngx_DEBUG, "sent config update to node")
+          end
         end
-      end
 
-    elseif err ~= "timeout" then
-      ngx_log(ngx_ERR, "semaphore wait error: ", err)
+      elseif err ~= "timeout" then
+        return nil, "semaphore wait error: " .. err
+      end
     end
+  end)
+
+  local _, err = ngx.thread.wait(main_thread, ping_thread)
+
+  ngx.thread.kill(main_thread)
+  ngx.thread.kill(ping_thread)
+
+  wb:send_close()
```

Comment:
```suggestion
  wb:close()
```

We should just close the WS connection here, if `send_close` hasn't been called, then the `close` method will call it automatically. Because the other end might be dead by now, and `send_close` may keep the (dead) connection around for longer than we need.